### PR TITLE
New collective

### PR DESCRIPTION
**Pull request by ipdemes**

### CODE REVIEW

File: src/core/mapping/base_mapper.cc
Comment (magnatelee) [inline src/core/mapping/base_mapper.cc:231]:
```suggestion
  for (uint32_t idx = 0; idx < task.regions.size(); ++idx) {
    auto& req = task.regions[idx];
    if (find_legate_projection_functor(req.projection)->is_collective())
      output.check_collective_regions.insert(idx);
  }
```

File: src/core/runtime/projection.cc
Comment (magnatelee) [inline src/core/runtime/projection.cc:128]: I assume `set_collective` is monotonic, so you can just return after calling it.

File: src/core/runtime/projection.cc
Comment (magnatelee) [inline src/core/runtime/projection.cc:135]: same here. you don't need to continue the check if you already set the flag.

File: src/core/mapping/base_mapper.cc
Comment (ipdemes) [reply src/core/mapping/base_mapper.cc:231]: @magnatelee :did you remove 214-216, 220-222  intentionally? Without it reduction and broadcasted instances are not marked as collective.

Comment (magnatelee) [reply src/core/mapping/base_mapper.cc]: I forgot a condition in the if statement. Here's the right one:
```
for (uint32_t idx = 0; idx < task.regions.size(); ++idx) {
    auto& req = task.regions[idx];
    if (req.handle_type == LEGION_SINGULAR_PROJECTION || find_legate_projection_functor(req.projection)->is_collective())
      output.check_collective_regions.insert(idx);
  }
```
Note that your original code is incorrect because of two reasons: 1) it mis-categorizes a read-only requirement with an identity projection as a broadcasting requirement, 2) it misses a reduction requirement on a region, because that requirement has 0 for its `projection`.

File: src/core/runtime/projection.h
Comment (magnatelee) [inline src/core/runtime/projection.h:49]: a nitpick: suffixing, not prefixing, member variables with `_` has been the convention.

File: src/core/mapping/base_mapper.cc
Comment (magnatelee) [inline src/core/mapping/base_mapper.cc:976]: remove these bogus changes

Comment (ipdemes) [reply src/core/mapping/base_mapper.cc:976]: the ` assert(!band_ranking.empty());` doesn't work anymore when all instances are collective. Do you think I need to modify logic for legate_select_sources` to parse thought collectiveViews as well?

File: src/core/runtime/projection.h
Comment (magnatelee) [inline src/core/runtime/projection.h:41]: this could be a `const` method

**Comment by magnatelee** [comment]
Looks like this PR contains changes from #511, so I hold off on reviewing this until #511 is merged.

File: src/core/mapping/base_mapper.cc
Comment (magnatelee) [inline src/core/mapping/base_mapper.cc:220]:
```suggestion
  for (uint32_t idx = 0; idx < task.regions.size(); ++idx) {
    auto& req   = task.regions[idx];
    if ((req.handle_type == LEGION_SINGULAR_PROJECTION) ||
        (find_legate_projection_functor(req.projection)->is_collective())) {
      output.check_collective_regions.insert(idx);
    }
  }

```

Comment (magnatelee) [reply src/core/mapping/base_mapper.cc:220]: fyi, `find_legate_projection_functor` can't return a `nullptr` anymore.

**Review by magnatelee — APPROVED** [review]

File: src/core/mapping/base_mapper.cc
Comment (magnatelee) [inline src/core/mapping/base_mapper.cc:211]: Let's wrap this loop with a C macro, say `LEGATE_NO_COLLECTIVE_INSTANCE`, so we can build the code with the `control_replication` branch as well. in the worst case where we don't do perf testing with the `collective` branch, we will need to create packages without collective instances.

Comment (ipdemes) [reply src/core/mapping/base_mapper.cc:211]: sounds good

**Comment by manopapad** [comment]
@ipdemes I think `legate_select_sources` now needs to consider `input.collective_views` as well. According to @lightsighter , `PhysicalInstance`s that are part of a collective view will appear only in `collective_views`, and will not appear in `source_instances`.

**Pull request merged**
